File: Max_Dot_Product_of_Two_Subsequences.java
class Solution {
    public int maxDotProduct(int[] nums1, int[] nums2) {
        int m = nums1.length;
        int n = nums2.length;

        if (m < n) {
            return maxDotProduct(nums2, nums1);
        }

        long[] dp = new long[n + 1];
        Arrays.fill(dp, Integer.MIN_VALUE);

        for (int i = 0; i < m; i++) {
            long prev = 0;
            for (int j = 0; j < n; j++) {
                long tmp = dp[j + 1];
                dp[j + 1] = Math.max(prev + nums1[i] * nums2[j], Math.max(nums1[i] * nums2[j], Math.max(dp[j], dp[j + 1])));
                prev = tmp;
            }
        }

        return (int)dp[n];
    }
}

File: Max_Dot_Product_of_Two_Subsequences.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var maxDotProduct = function(nums1, nums2) {
        const m = nums1.length;
        const n = nums2.length;

        if (m < n) {
            return maxDotProduct(nums2, nums1);
        }

        const dp = new Array(n + 1).fill(Number.NEGATIVE_INFINITY);

        for (let i = 0; i < m; i++) {
            let prev = 0;
            for (let j = 0; j < n; j++) {
                const tmp = dp[j + 1];
                dp[j + 1] = Math.max(prev + nums1[i] * nums2[j], nums1[i] * nums2[j], dp[j], dp[j + 1]);
                prev = tmp;
            }
        }

        return dp[n];    
};
